pkg/alert/alert.go

```diff
@@ -0,0 +1,327 @@
+	logger        log.Logger
+	alertmanagers func() []*url.URL
+	doReq         func(req *http.Request) (*http.Response, error)
+
+	sent    *prometheus.CounterVec
+	dropped *prometheus.CounterVec
+	latency *prometheus.HistogramVec
+}
+
+// NewSender returns a new sender. On each call to Send the entire alert batch is sent
+// to each Alertmanager returned by the getter function.
+func NewSender(
+	logger log.Logger,
+	reg prometheus.Registerer,
+	alertmanagers func() []*url.URL,
+	doReq func(req *http.Request) (*http.Response, error),
+) *Sender {
+	if doReq == nil {
+		doReq = http.DefaultClient.Do
+	}
+	if logger == nil {
+		logger = log.NewNopLogger()
+	}
+	s := &Sender{
+		logger:        logger,
+		alertmanagers: alertmanagers,
+		doReq:         doReq,
+
+		sent: prometheus.NewCounterVec(prometheus.CounterOpts{
+			Name: "thanos_alert_sender_alerts_sent_total",
+			Help: "Total number of alerts sent by alertmanager.",
+		}, []string{"alertmanager"}),
+
+		dropped: prometheus.NewCounterVec(prometheus.CounterOpts{
+			Name: "thanos_alert_sender_alerts_dropped_total",
+			Help: "Total number of alerts dropped by alertmanager.",
+		}, []string{"alertmanager"}),
+
+		latency: prometheus.NewHistogramVec(prometheus.HistogramOpts{
+			Name: "thanos_alert_sender_latency_seconds",
+			Help: "Latency for sending alert notifications (not including dropped notifications).",
+		}, []string{"alertmanager"}),
+	}
+	if reg != nil {
+		reg.MustRegister(s.sent, s.dropped, s.latency)
+	}
+	return s
+}
+
+// Send an alert batch to all given Alertmanager URLs.
+func (s *Sender) Send(ctx context.Context, alerts []*Alert) error {
+	if len(alerts) == 0 {
+		return nil
+	}
+	b, err := json.Marshal(alerts)
+	if err != nil {
+		return errors.Wrap(err, "encode alerts")
+	}
+
+	var g errgroup.Group
+
+	for _, u := range s.alertmanagers() {
+		sendCtx, cancel := context.WithCancel(ctx)
+
+		g.Go(func() error {
+			defer cancel()
+
+			start := time.Now()
+			u.Path = path.Join(u.Path, alertPushEndpoint)
+
+			if err := s.sendOne(sendCtx, u.String(), b); err != nil {
+				level.Warn(s.logger).Log(
+					"msg", "sending alerts failed",
+					"alertmanager", u.Host,
+					"numDropped", len(alerts))
+				s.dropped.WithLabelValues(u.Host).Add(float64(len(alerts)))
+				return err
+			}
+			s.sent.WithLabelValues(u.Host).Add(float64(len(alerts)))
+			s.latency.WithLabelValues(u.Host).Observe(time.Since(start).Seconds())
+			return nil
+		})
+	}
+	if err := g.Wait(); err != nil {
+		return errors.Wrap(err, "send alerts")
+	}
+	return nil
+}
+
+func (s *Sender) sendOne(ctx context.Context, url string, b []byte) error {
+	req, err := http.NewRequest("POST", url, bytes.NewReader(b))
+	if err != nil {
+		return err
+	}
+	req = req.WithContext(ctx)
+	req.Header.Set("Content-Type", contentTypeJSON)
+
+	resp, err := s.doReq(req)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode/100 != 2 {
+		return errors.Errorf("bad response status %v", resp.Status)
```

Comment:
+ url

Reply:
?

Reply:
I wrote `+Url` but github translated that as bullet point ;p